cranelift-filetests/src/test_binemit.rs

```diff
@@ -337,3 +316,41 @@ impl SubTest for TestBinEmit {
         Ok(())
     }
 }
+
+/// validate registers/stack slots are correctly annotated
+fn validate_annotation(
+    func: &ir::Function,
+    inst: ir::Inst,
+    isa: &dyn isa::TargetIsa,
+    validate_args: bool,
+) -> SubtestResult<()> {
+    if let Some(&v) = func
+        .dfg
+        .inst_results(inst)
+        .iter()
+        .find(|&&v| !func.locations[v].is_assigned())
+    {
+        Err(format!(
+            "Need register/stack slot annotation for {} in {}",
+            v,
+            func.dfg.display_inst(inst, isa)
+        ))
+    } else if validate_args {
+        if let Some(&v) = func
+            .dfg
+            .inst_args(inst)
```

Comment:
Alternatively, to reduce code duplication, could the function either look at args *or* results, but not both at the same time? In the second call site, it would be called twice instead of once, which is fine.

Then you can do something like `let values = if validate_args { func.dfg.inst_args(inst) } else { func.dfg.inst_results(inst) };` and check the values' locations.